### ISSUE TITLE
Unfold tree (unfold everything under definition) feature

### ISSUE THREAD

**Issue by siddharth-krishna** [atom/atom#11785]
When I open a file, I first use the `Editor: Fold All` command to get a high-level view of the file, and then locate the function I want to look into. Now, to actually look at the code of that function I unfold the line where the function begins by clicking the little `>` arrow to it's left.

The problem is when functions contain other functions, then all the sub-functions are by default folded too. So to see the full code of the sub-function, I need to manually click and unfold each of them, and possibly the sub-sub-functions too. In a language like OCaml, where it's common coding practice to have many nested `let ... in` structures, this makes it a huge pain to use the folding feature.

**Suggested enhancement**

Instead of toggling between fold and unfold when clicking the arrow, instead cycle through "fold", "unfold", and "unfold entire tree".

**Current and suggested behavior**

Currently, after a fold-all, my code looks like this:

```
> def foo(n): ...
```
If I click the arrow once, I get:
```
 def foo(n):
>    def bar():...
>    def boo():...
```
And to see everything I need to click and unfold `bar` and `boo` manually. Clicking the arrow next to `foo` once more just returns it to the folded state in the first code block above.

Instead, I suggest that clicking the `foo` the second time produces this instead:
```
 def foo(n):
     def bar():
         pass
     def boo():
         pass
```

**Why would the enhancement be useful to most users**

It would save users of languages like Ocaml many clicks, and they'd be able to easily see the entire code of a function.

This feature exists in Emacs, for example, the default behaviour of <Tab> folding in org-mode does exactly this.

**Atom Version:** 1.7.3
**OS and Version:** OS X 10.10.5

**Comment by mathieufrh** [comment]
Currently in Atom 1.11.1 on linux, when a portion of code is folded it looks like:

```
Currency::Rebase() {=
```

If I double-click on the row, after the `=` symbol then it unfold the row and all the underlying folded rows.

**Comment by siddharth-krishna** [comment]
Hmm, that doesn't work for me. I'm on OS X, Atom version 1.11.2..

**Comment by sarimak** [comment]
Also doesn't work for me - Linux, Atom 1.11.2, Python syntax.
Clicking on the ellipsis at the end of folded line unfolds only a direct child, not whole sub-tree.

**Comment by ahwebd** [comment]
`Editor: Fold All` will fold all levels (1st level and its nested levels) so don't use it at the beginning, instead use: `Editor: Fold At Indent Level 1` (`Ctrl+k Ctrl+1`) to fold only the 1st level (nested levels will still be unfolded inside) and then when you click to unfold a block all code inside the block will be expanded.

**Comment by jrial** [comment]
@ahwebd Good suggestion, thanks! Although I'd prefer to have Fold All do the logical thing (what "Fold to Level 1" does), rather than "Hierarchically Fold All Nesting Levels From Deepest All The Way To Level One", which I can't imagine any practical use for.

It would also be nice if the folding shortcuts would only operate on the selection when text is selected, but that's a separate issue.

**Comment by Jay-Squared** [comment]
Fun fact: I came to this site by searching how exactly I could hierarchically fold all nesting levels from the lower all the way to level one. I didn't realise “Fold All” would do that. The practical use for me is that it gets very messy when I unfold every item at every level. I need all lower levels to stay in folded state to be able to navigate easily between the different sections.

**Comment by jrial** [comment]
Still happening. Atom 1.40.0

**Comment by bmundt6** [comment]
Same issue here, MacOS 10.15.2 and Atom 1.42.0

**Comment by Asday** [comment]
> `Editor: Fold All` will fold all levels (1st level and its nested levels) so don't use it at the beginning, instead use: `Editor: Fold At Indent Level 1` (`Ctrl+k Ctrl+1`) to fold only the 1st level (nested levels will still be unfolded inside) and then when you click to unfold a block all code inside the block will be expanded.

It's not a great solution.

```python3
class A:

    def b(self):
        ...

    def c(self):
        ...
```

Folding at the top level only means when I unfold `A`, the entirety of `b` and `c` are shown.  There are times where I want to know everything `A` has, which is achievable if I first `Ctrl+Shift+Alt+[`, but then I'd like to subsequently descend into one of `A`'s members, and probably see its entire contents, but sometimes not.  Examples are if I want to read the implementation of a function for the former, or just the docstring for the latter.

I might then want to unfold something within _there_ fully, say there's a function or class defined within that function or class.

Put simply, I want a "fully unfold current line".
